### DISCUSSION TITLE
DataLoader is called several times on a mutation request.

### DISCUSSION THREAD

**Discussion by ngsw-taro** [Netflix/dgs-framework#969] [Q&A]
I'll omit unimportant parts for simplicity but I throw the following request to create multiple books.

```gql
mutation {
  createBooks {
    authors {
      name
    }
  }
}
```

Then the following code runs and `AuthorsLoader#load` is called multiple times.
As a result I have the N + 1 problem.
As you know, this problem doesn't occurred on query requests.

```kotlin
@DgsData(parentType = "Book")
fun authors(e: DgsDataFetchingEnvironment): CompletableFuture<List<Author>> {
    val book = e.getSource<Book>()
    val loader = e.getDataLoader<String, List<Author>>("authors")
    return loader.load(book.id)
}

@DgsDataLoader(name = "authors")
class AuthorsLoader : MappedBatchLoader<String, List<Author>> {
    override fun load(keys: Set<String>): CompletionStage<Map<String, List<Author>>> {
        /* This method called 2 times or more */
        return CompletableFuture.supplyAsync { ... }
    }
}
```

How do I resolve?
Would be great if someone can help.
Thanks

**Comment by paulbakker** [comment]
So let's say you load book 1 and book 2. You're seeing `load` being called twice, once for each book, and `keys` only contain a single id?

**Reply by ngsw-taro** [reply]
Thanks for your reply. You're absolutely right!

**Comment by Diego-SA** [comment]
I have found the same problem but with a tiny different approach. In my case, I have an object with two attributes, both of them containing a new object that must be treated by the dataloader. I will present a context similar to the discussion created. Imagine that you have a shareBook mutation where you indicate a book to give and a book to receive. It could have more fields but I will simplify it.

```
shareBook {
   giveBook {
      book
   }
   receiveBook {
      book
   }
}
```
In this case, when we use a BookDataLoader, it will be called twice if it is a mutation (e.g. create a shareBook registry), but it will be called once if it is a query (e.g. find a shareBook object by ID).

I think it's a library error, but it would be great if someone can propose a solution or at least an explanation about why mutation is being processed differently.

Thanks.
